codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/PubCrateConstrainedShapeSymbolProvider.kt

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.rust.codegen.client.smithy
+
+import software.amazon.smithy.codegen.core.Symbol
+import software.amazon.smithy.model.Model
+import software.amazon.smithy.model.knowledge.NullableIndex
+import software.amazon.smithy.model.shapes.CollectionShape
+import software.amazon.smithy.model.shapes.MapShape
+import software.amazon.smithy.model.shapes.MemberShape
+import software.amazon.smithy.model.shapes.ServiceShape
+import software.amazon.smithy.model.shapes.Shape
+import software.amazon.smithy.model.shapes.SimpleShape
+import software.amazon.smithy.model.shapes.StructureShape
+import software.amazon.smithy.model.shapes.UnionShape
+import software.amazon.smithy.rust.codegen.client.rustlang.RustReservedWords
+import software.amazon.smithy.rust.codegen.client.rustlang.RustType
+import software.amazon.smithy.rust.codegen.core.util.PANIC
+import software.amazon.smithy.rust.codegen.core.util.toPascalCase
+import software.amazon.smithy.rust.codegen.core.util.toSnakeCase
+
+// TODO Move this to `core` or `server`.
+
+/**
+ * The [PubCrateConstrainedShapeSymbolProvider] returns, for a given
+ * _transitively but not directly_ constrained shape, a symbol whose Rust type
+ * can hold the constrained values.
+ *
+ * For collection and map shapes, this type is a [RustType.Opaque] wrapper
+ * tuple newtype holding a container over the inner constrained type. For
+ * member shapes, it's whatever their target shape resolves to.
+ *
+ * The class name is prefixed with `PubCrate` because the symbols it returns
+ * have associated types that are generated as `pub(crate)`. See the
+ * `PubCrate*Generator` classes to see how these types are generated.
+ *
+ * It is important that this symbol provider _not_ wrap
+ * [ConstrainedShapeSymbolProvider], since otherwise it will eventually
+ * delegate to it and generate a symbol with a `pub` type.
+ *
+ * Note simple shapes cannot be transitively but not directly constrained, so
+ * this symbol provider is only implemented for aggregate shapes. The symbol
+ * provider will intentionally crash in such a case to avoid the caller
+ * incorrectly using it.
```

Comment:
I meant to convey that simple shapes cannot be transitively constrained, but not directly, since they're... simple. I've reworded this in https://github.com/awslabs/smithy-rs/pull/1342/commits/881ecb3f11fcdd4b5f9e09bdd02fff147476ebc5.